src/bootstrap/tool.rs

```diff
@@ -514,6 +514,19 @@ impl Step for Rustdoc {
         // rustc compiler it's paired with, so it must be built with the previous stage compiler.
         let build_compiler = builder.compiler(target_compiler.stage - 1, builder.config.build);
 
+        // When using `download-rustc` and a stage0 build_compiler, copying rustc doesn't actually
+        // build stage0 libstd (because the libstd in sysroot has the wrong ABI). Explicitly build
+        // it.
```

Comment:
Hm, this seems suspicious - it feels like we should probably fix assemble or w/e to build it? Why are we explicitly building it here in impl step for rustdoc rather than fixing the origin of the problem (where we skip the build)?

Reply:
Hmm, in theory I guess you're right, but if building the stage1 sysroot required building the compiler each time it defeats the whole point of `download-rustc`, since building stage2 rustdoc requires having a stage1 sysroot.

This does have the downside that each tool has to add its own `ensure()` call, but I don't know if that's a problem in practice - like I said below, I don't think there's much use in compiling with beta instead of with master.